yarn-project/entrypoints/src/payload.ts

```diff
@@ -3,117 +3,107 @@ import { padArrayEnd } from '@aztec/foundation/collection';
-export abstract class EntrypointPayload {
-  protected constructor(
-    private functionCalls: EncodedFunctionCall[],
-    private _hashedArguments: HashedValues[],
-    private generatorIndex: number,
-    private _nonce: Fr,
+/** Represents data necessary to execute a list of function calls successfully */
+export class ExecutionPayload {
+  public constructor(
+    /** The function calls to be executed. */
+    public calls: FunctionCall[],
+    /** Any transient auth witnesses needed for this execution */
+    public authWitnesses: AuthWitness[],
+    /** Data passed through an oracle for this execution. */
+    public capsules: Capsule[],
+    /** Extra hashed values to be injected in the execution cache */
+    public extraHashedValues: HashedValues[] = [],
   ) {}
 
-  protected static async create(functionCalls: FunctionCall[]) {
+  static empty() {
+    return new ExecutionPayload([], [], []);
+  }
+
+  /**
+   * Encodes the payload for execution, following Noir's convention
+   */
+  public async encode(): Promise<EncodedExecutionPayload> {
     const hashedArguments: HashedValues[] = [];
-    for (const call of functionCalls) {
+    for (const call of this.calls) {
       hashedArguments.push(await HashedValues.fromValues(call.args));
     }
 
     /* eslint-disable camelcase */
-    const encodedFunctionCalls = functionCalls.map((call, index) => ({
+    const encodedFunctionCalls: EncodedFunctionCall[] = this.calls.map((call, index) => ({
       args_hash: hashedArguments[index].hash,
       function_selector: call.selector.toField(),
       target_address: call.to.toField(),
       is_public: call.type == FunctionType.PUBLIC,
       is_static: call.isStatic,
     }));
-    /* eslint-enable camelcase */
 
     return {
       encodedFunctionCalls,
-      hashedArguments,
+      hashedArguments: [...hashedArguments, ...this.extraHashedValues],
+      authWitnesses: this.authWitnesses,
+      capsules: this.capsules,
+      function_calls: encodedFunctionCalls,
     };
+    /* eslint-enable camelcase */
   }
+}
 
+/**
+ * Representation of the encoded payload for execution
+ */
+export type EncodedExecutionPayload = Omit<ExecutionPayload, 'calls' | 'encode' | 'extraHashedValues'> & {
+  /** Function calls in the expected format (Noir's convention) */
+  encodedFunctionCalls: EncodedFunctionCall[];
+  /** The hashed args for the call, ready to be injected in the execution cache */
+  hashedArguments: HashedValues[];
   /* eslint-disable camelcase */
   /**
    * The function calls to execute. This uses snake_case naming so that it is compatible with Noir encoding
-   * @internal
-   */
-  get function_calls() {
-    return this.functionCalls;
-  }
+   * */
+  get function_calls(): EncodedFunctionCall[];
   /* eslint-enable camelcase */
+};
 
-  /**
-   * The nonce
-   * @internal
-   */
-  get nonce() {
-    return this._nonce;
-  }
+/** Represents the ExecutionPayload after encoding for the entrypint to execute */
```

Comment:
```suggestion
/** Represents the ExecutionPayload after encoding for the entrypont to execute */
```
Could you expand on what does it mean to encode for an entrypoint here?

Reply:
Unfortunately it's a sad situation. Entrypoints require function calls to be encoded in noir format (snake_case, args hash, etc). The whole payload shouldn't have to be encoded, but changing the model turned out to be extremely complicated in this PR. Will change soon